tests/unit/s2n_tls13_cert_verify_test.c

```diff
@@ -234,5 +234,122 @@ int main(int argc, char **argv)
         run_tests(&test_cases[i], S2N_SERVER);
     }
 
+    /* Self-talk: Ensure that the signature algorithm used to sign the CertificateVerify message
+     * is validated against the certificate type
+     */
+    if (s2n_is_tls13_fully_supported()) {
+        struct s2n_tls13_cert_verify_test test_server_parameters[] = {
+            {
+                    .cert_file = S2N_RSA_2048_PKCS1_CERT_CHAIN,
+                    .key_file = S2N_RSA_2048_PKCS1_KEY,
+                    .sig_scheme = &s2n_rsa_pss_rsae_sha256,
+            },
+            {
+                    .cert_file = S2N_RSA_PSS_2048_SHA256_LEAF_CERT,
+                    .key_file = S2N_RSA_PSS_2048_SHA256_LEAF_KEY,
+                    .sig_scheme = &s2n_rsa_pss_pss_sha256,
+            },
+            {
+                    .cert_file = S2N_ECDSA_P256_PKCS1_CERT_CHAIN,
+                    .key_file = S2N_ECDSA_P256_PKCS1_KEY,
+                    .sig_scheme = &s2n_ecdsa_sha256,
+            }
+        };
+
+        const struct s2n_signature_scheme *test_client_sig_schemes[] = {
+            &s2n_rsa_pss_rsae_sha256,
+            &s2n_rsa_pss_pss_sha256,
+            &s2n_ecdsa_sha256,
+        };
+
+        for (size_t param_idx = 0; param_idx < s2n_array_len(test_server_parameters); param_idx++) {
+            struct s2n_tls13_cert_verify_test server_parameters = test_server_parameters[param_idx];
+
+            DEFER_CLEANUP(struct s2n_cert_chain_and_key *cert_chain = NULL, s2n_cert_chain_and_key_ptr_free);
+            EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&cert_chain, server_parameters.cert_file,
+                    server_parameters.key_file));
+
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "test_all_tls13"));
+            EXPECT_SUCCESS(s2n_config_disable_x509_verification(config));
+
+            /* The server only supports the single test certificate, and will only use the
+             * corresponding signature algorithm to sign the CertificateVerify message, regardless
+             * of the client's advertised signature schemes.
+             */
```

Comment:
If I'm following, this test is relying on our sig alg selection logic having fallback logic in the case of no shared sig alg options? So this would be broken if we considered no matches an error instead of just trying an unsupported sig alg anyway? It's probably a good idea to specifically call out that you're relying on that behavior, because that behavior could change someday. It's not exactly RFC compliant or popular :)